Thread local debug info

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22365 Thread local debug info**

Summary:
We'd want to pass debug information (e.g. model_id) from the model
invocation site down to RecordFunction debug/logging callbacks without
necessarily changing an interface for each new field
This PR adds a thread local debug info storage and makes sure we keep
the information across inter-op calls

Test Plan:
USE_CUDA=0 python setup.py develop
./build/bin/test_jit

Differential Revision: [D16065129](https://our.internmc.facebook.com/intern/diff/D16065129)

Commit:
[wip] Thread local debug info

Summary:
We'd want to pass debug information (e.g. model_id) from the model
invocation site down to RecordFunction debug/logging callbacks without
necessarily changing an interface for each new field
This PR adds a thread local debug info storage and makes sure we keep
the information across inter-op calls

Test Plan:
USE_CUDA=0 python setup.py develop
./build/bin/test_jit

ilia-cher commented:
we also need to pass debug info into backward pass tasks

Commit:
Update on "[wip] Thread local debug info"

[wip] Thread local debug info

Summary:
We'd want to pass debug information (e.g. model_id) from the model
invocation site down to RecordFunction debug/logging callbacks without
necessarily changing an interface for each new field
This PR adds a thread local debug info storage and makes sure we keep
the information across inter-op calls

Test Plan:
USE_CUDA=0 python setup.py develop
./build/bin/test_jit

gh-metadata: pytorch pytorch 22365 gh/ilia-cher/35/head

Commit:
Update on "[wip] Thread local debug info"

[wip] Thread local debug info

Summary:
We'd want to pass debug information (e.g. model_id) from the model
invocation site down to RecordFunction debug/logging callbacks without
necessarily changing an interface for each new field
This PR adds a thread local debug info storage and makes sure we keep
the information across inter-op calls

Test Plan:
USE_CUDA=0 python setup.py develop
./build/bin/test_jit

gh-metadata: pytorch pytorch 22365 gh/ilia-cher/35/head

Commit:
Update on "[wip] Thread local debug info"

[wip] Thread local debug info

Summary:
We'd want to pass debug information (e.g. model_id) from the model
invocation site down to RecordFunction debug/logging callbacks without
necessarily changing an interface for each new field
This PR adds a thread local debug info storage and makes sure we keep
the information across inter-op calls

Test Plan:
USE_CUDA=0 python setup.py develop
./build/bin/test_jit

gh-metadata: pytorch pytorch 22365 gh/ilia-cher/35/head

Commit:
Update on "[wip] Thread local debug info"

[wip] Thread local debug info

Summary:
We'd want to pass debug information (e.g. model_id) from the model
invocation site down to RecordFunction debug/logging callbacks without
necessarily changing an interface for each new field
This PR adds a thread local debug info storage and makes sure we keep
the information across inter-op calls

Test Plan:
USE_CUDA=0 python setup.py develop
./build/bin/test_jit

gh-metadata: pytorch pytorch 22365 gh/ilia-cher/35/head

Commit:
Update on "[wip] Thread local debug info"

[wip] Thread local debug info

Summary:
We'd want to pass debug information (e.g. model_id) from the model
invocation site down to RecordFunction debug/logging callbacks without
necessarily changing an interface for each new field
This PR adds a thread local debug info storage and makes sure we keep
the information across inter-op calls

Test Plan:
USE_CUDA=0 python setup.py develop
./build/bin/test_jit

gh-metadata: pytorch pytorch 22365 gh/ilia-cher/35/head

Commit:
Update on "[wip] Thread local debug info"

[wip] Thread local debug info

Summary:
We'd want to pass debug information (e.g. model_id) from the model
invocation site down to RecordFunction debug/logging callbacks without
necessarily changing an interface for each new field
This PR adds a thread local debug info storage and makes sure we keep
the information across inter-op calls

Test Plan:
USE_CUDA=0 python setup.py develop
./build/bin/test_jit

gh-metadata: pytorch pytorch 22365 gh/ilia-cher/35/head

dzhulgakov commented on aten/src/ATen/ThreadLocalDebugInfo.h:
```diff
@@ -0,0 +1,24 @@
+  ThreadLocalDebugInfoBase() {}
+  virtual ~ThreadLocalDebugInfoBase() {}
+
+  virtual std::string getString(const char* name) = 0;
```
we could also just do dynamic_cast to whatever type needed, but it works too

apaszke replied:
Why do we require implementing those two methods in particular? What's so special about them?

dzhulgakov commented on aten/src/ATen/ThreadLocalDebugInfo.cpp:
```diff
@@ -0,0 +1,18 @@
+
+void setThreadLocalDebugInfo(
+    std::shared_ptr<ThreadLocalDebugInfoBase> info) {
+  debug_info = info;
```
nit: std::move

dzhulgakov commented on aten/src/ATen/ParallelThreadPoolNative.cpp:
```diff
@@ -67,10 +68,17 @@ int get_num_interop_threads() {
+  auto debug_info = getThreadLocalDebugInfo();
+  auto fn = [func, debug_info]() {
+    auto prev_info = getThreadLocalDebugInfo();
+    setThreadLocalDebugInfo(debug_info);
```
nit: std::move?

also you can make this API to return the previous version (do a swap)

dzhulgakov reviewed (commented):
Looks good. I think we should instrument autograd too (including reentrant backwards) - probably somewhere here - https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/engine.cpp

apaszke reviewed (changes requested):
What's the plan for this feature? If it's supposed to reliably propagate through all the places where it's us controlling the threads and not the user, why aren't we e.g. taking care of it for backwards? It doesn't seem fully fleshed out, especially with the weird `ThreadLocalDebugInfoBase` API.

dzhulgakov commented on aten/src/ATen/ThreadLocalDebugInfo.h:
```diff
@@ -0,0 +1,24 @@
+
+namespace at {
+
+class CAFFE2_API ThreadLocalDebugInfoBase {
```
Can't we just use shared_ptr<void> for it? If we wanted more safety, we could use our own RTTI (based on TypeMeta) but it'd be a bit weird use case for it.

ilia-cher commented:
> What's the plan for this feature? If it's supposed to reliably propagate through all the places where it's us controlling the threads and not the user, why aren't we e.g. taking care of it for backwards? It doesn't seem fully fleshed out, especially with the weird `ThreadLocalDebugInfoBase` API.

I've commented "we also need to pass debug info into backward pass tasks" above and also put "[wip]" - "work in progress" in the title, we do need to take care for backwards pass too.

Commit:
Update on "[wip] Thread local debug info"


Summary:
We'd want to pass debug information (e.g. model_id) from the model
invocation site down to RecordFunction debug/logging callbacks without
necessarily changing an interface for each new field
This PR adds a thread local debug info storage and makes sure we keep
the information across inter-op calls

Test Plan:
USE_CUDA=0 python setup.py develop
./build/bin/test_jit

Differential Revision: [D16065129](https://our.internmc.facebook.com/intern/diff/D16065129)

Commit:
Update on "Thread local debug info"


Summary:
We'd want to pass debug information (e.g. model_id) from the model
invocation site down to RecordFunction debug/logging callbacks without
necessarily changing an interface for each new field
This PR adds a thread local debug info storage and makes sure we keep
the information across inter-op calls

Test Plan:
USE_CUDA=0 python setup.py develop
./build/bin/test_jit

Differential Revision: [D16065129](https://our.internmc.facebook.com/intern/diff/D16065129)

Commit:
Update on "Thread local debug info"


Summary:
We'd want to pass debug information (e.g. model_id) from the model
invocation site down to RecordFunction debug/logging callbacks without
necessarily changing an interface for each new field
This PR adds a thread local debug info storage and makes sure we keep
the information across inter-op calls

Test Plan:
USE_CUDA=0 python setup.py develop
./build/bin/test_jit

Differential Revision: [D16065129](https://our.internmc.facebook.com/intern/diff/D16065129)

Commit:
Update on "Thread local debug info"


Summary:
We'd want to pass debug information (e.g. model_id) from the model
invocation site down to RecordFunction debug/logging callbacks without
necessarily changing an interface for each new field
This PR adds a thread local debug info storage and makes sure we keep
the information across inter-op calls

Test Plan:
USE_CUDA=0 python setup.py develop
./build/bin/test_jit

Differential Revision: [D16065129](https://our.internmc.facebook.com/intern/diff/D16065129)

ilia-cher commented:
Addressed comments, added debug info into backward pass, improved test - see `testThreadLocalDebugInfo` as an example of a use case

dzhulgakov commented on aten/src/ATen/ThreadLocalDebugInfo.h:
```diff
@@ -0,0 +1,22 @@
+  virtual ~ThreadLocalDebugInfoBase() {}
+};
+
+CAFFE2_API std::shared_ptr<ThreadLocalDebugInfoBase> getThreadLocalDebugInfo();
```
maybe add a short comment about all this functionality?

dzhulgakov reviewed (approved):
Can you also please extend the note with documentation: https://pytorch.org/docs/master/notes/large_scale_deployments.html

Otherwise - looks good to me (@apaszke @ezyang - feel free to chime in with autograd comments)

ezyang commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -657,6 +661,7 @@ auto Engine::execute(const edge_list& roots,
   ClearCallbacks _cb_guard(final_callbacks_, post_callbacks_lock_);
 
   GraphTask graph_task(keep_graph, create_graph, worker_device == NO_DEVICE ? 0 : total_depth+1);
+  graph_task.debug_info_ = at::getThreadLocalDebugInfo();
```
Why not set this up in the constructor?

ezyang commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -495,9 +498,10 @@ static variable_list call_function(NodeTask& task) {
+  } else {
     outputs = fn(std::move(inputs));
   }
+  at::setThreadLocalDebugInfo(prev_info);
```
This doesn't look exception safe.

ilia-cher replied:
I'll mark the API as noexcept

apaszke replied:
As in use RAII

ilia-cher replied:
sounds good, let me add a guard object to use

ezyang commented on aten/src/ATen/ParallelThreadPoolNative.cpp:
```diff
@@ -67,10 +68,16 @@ int get_num_interop_threads() {
+  auto fn = [func, debug_info]() {
+    auto prev_info = setThreadLocalDebugInfo(std::move(debug_info));
+    func();
+    setThreadLocalDebugInfo(std::move(prev_info));
```
This doesn't look exception safe either.

ilia-cher replied:
is it the shared_ptr move the concern here?

dzhulgakov replied:
I guess now it's fine with noexcept. I'm curious about general exception handling in at::parallel. Do we rely on `func` to never throw? (otherwise it'd lead to hard-to-debug bugs). In any case - would be nice to add a comment about exception handling (it's a common pitfall - I can easily recall 3-4 times when we messed it up)

ilia-cher replied:
@dzhulgakov, we have three implementations of parallel_for (OpenMP, TBB and native one) and all of them capture exceptions from the given functor (from within the subtasks) and then rethrow the exception outside of parallel_for

ezyang replied:
I was referring to the `func()` call. If the invariant on `launch` is that the passed in functor should never throw, that should be documented on the function.

ezyang reviewed (changes requested):
Exception safety problems.

Commit:
Update on "Thread local debug info"


Summary:
We'd want to pass debug information (e.g. model_id) from the model
invocation site down to RecordFunction debug/logging callbacks without
necessarily changing an interface for each new field
This PR adds a thread local debug info storage and makes sure we keep
the information across inter-op calls

Test Plan:
USE_CUDA=0 python setup.py develop
./build/bin/test_jit

Differential Revision: [D16065129](https://our.internmc.facebook.com/intern/diff/D16065129)

ilia-cher commented:
Addressed comments,
@ezyang is the concern about moving shared_ptr (re. exception safety)? I marked the functions as noexcept, I don't expect this to throw but if it does then it will terminate

dzhulgakov reviewed (approved):
If tests pass - it's good to go. Thanks for the docs!

apaszke commented on aten/src/ATen/ThreadLocalDebugInfo.cpp:
```diff
@@ -0,0 +1,20 @@
+  auto ret = std::move(debug_info);
+  debug_info = std::move(info);
+  return ret;
+}
```
Why can't we make this an inline function? You're calling this from a hot-spot in autograd, so it would be nice to get it inlined.

apaszke reviewed (commented):
I'm a little concerned about stuffing more things into the autograd hotpath 😕

ilia-cher commented:
> I'm a little concerned about stuffing more things into the autograd hotpath 😕

I understand the concern, I also think adding more things into the hotpath should not be the norm, though in this case the extra benefit of logging and relative lightness (compared to all the rest of aurograd machinery) of setting of the debug info seem to outweight extra complexity in my opinion.
We should pay extra attention to the hotpath in any case though, to not allow it to slowly regress, I think continuous benchmarking would be right answer to this.

ezyang commented on aten/src/ATen/ThreadLocalDebugInfo.h:
```diff
@@ -0,0 +1,26 @@
+ public:
+  ThreadLocalDebugInfoBase() {}
+  virtual ~ThreadLocalDebugInfoBase() {}
+};
```
Non-blocking for this PR, but something to think about: the single thread local field that you've installed here is not compositional: if one client has DebugInfoA and another has DebugInfoB, there is no way for both of them to actually propagate both of them. Because of this lack of compositionality, I don't really see the point of abstracting the class in this way: it would probably be better to just have a single monolithic class that you put all of the stuff you want in, and if it becomes unmanageable at some later point in time, *then* add whatever abstraction you need to handle the debug information in a more scalable way. Basically, this is not one of those cases where generalizing early saves you pain later, so let's generalize as late as possible!

ezyang replied:
Unless, I suppose, what's actually happening is you're extending this class from inside fbcode, because the debug info you want to store doesn't exist inside PyTorch proper. But this isn't mentioned anywhere inside the PR; I'm just guessing ;)

ezyang replied:
I guess this is implied by the location of the documentation you added in docs/source/notes/large_scale_deployments.rst

ezyang replied:
I guess what I would like to see, is something in big, blinky, bold red letters, saying **DO NOT USE THIS IF YOU ARE A LIBRARY AUTHOR, THIS IS NOT FOR YOU**

ilia-cher replied:
@ezyang - yes, large_scale_deployments.rst explains the main use case and the test I added demoes it - we want to transparently pass something that does not exist within pure PT from the user to the user-authored hooks (e.g. logging). Not sure about big, blinky, bold red letters, but I'll add more into the comment in the code :)

ezyang reviewed (approved):
Okey dokey.

Commit:
Update on "Thread local debug info"


Summary:
We'd want to pass debug information (e.g. model_id) from the model
invocation site down to RecordFunction debug/logging callbacks without
necessarily changing an interface for each new field
This PR adds a thread local debug info storage and makes sure we keep
the information across inter-op calls

Test Plan:
USE_CUDA=0 python setup.py develop
./build/bin/test_jit

Differential Revision: [D16065129](https://our.internmc.facebook.com/intern/diff/D16065129)

Commit:
Update on "Thread local debug info"


Summary:
We'd want to pass debug information (e.g. model_id) from the model
invocation site down to RecordFunction debug/logging callbacks without
necessarily changing an interface for each new field
This PR adds a thread local debug info storage and makes sure we keep
the information across inter-op calls

Test Plan:
USE_CUDA=0 python setup.py develop
./build/bin/test_jit

Differential Revision: [D16065129](https://our.internmc.facebook.com/intern/diff/D16065129)

ilia-cher commented:
Adressed comments (RAII, inline, comment)

ilia-cher commented:
Ah, the windows build doesn't like dll export and thread_local at the same time

Commit:
Update on "Thread local debug info"


Summary:
We'd want to pass debug information (e.g. model_id) from the model
invocation site down to RecordFunction debug/logging callbacks without
necessarily changing an interface for each new field
This PR adds a thread local debug info storage and makes sure we keep
the information across inter-op calls

Test Plan:
USE_CUDA=0 python setup.py develop
./build/bin/test_jit

Differential Revision: [D16065129](https://our.internmc.facebook.com/intern/diff/D16065129)

Commit:
Update on "Thread local debug info"


Summary:
We'd want to pass debug information (e.g. model_id) from the model
invocation site down to RecordFunction debug/logging callbacks without
necessarily changing an interface for each new field
This PR adds a thread local debug info storage and makes sure we keep
the information across inter-op calls

Test Plan:
USE_CUDA=0 python setup.py develop
./build/bin/test_jit

Differential Revision: [D16065129](https://our.internmc.facebook.com/intern/diff/D16065129)

ilia-cher commented:
linter

Commit:
Update on "Thread local debug info"


Summary:
We'd want to pass debug information (e.g. model_id) from the model
invocation site down to RecordFunction debug/logging callbacks without
necessarily changing an interface for each new field
This PR adds a thread local debug info storage and makes sure we keep
the information across inter-op calls

Test Plan:
USE_CUDA=0 python setup.py develop
./build/bin/test_jit

Differential Revision: [D16065129](https://our.internmc.facebook.com/intern/diff/D16065129)

ezyang commented on aten/src/ATen/ParallelThreadPoolNative.cpp:
```diff
@@ -67,10 +68,19 @@ int get_num_interop_threads() {
 }
 
 void launch(std::function<void()> func) {
+  auto fn = std::bind([](
```
I'm surprised about the use of `std::bind` here. Normally a lambda is enough. What's going on here?

ilia-cher replied:
move capture is not supported till C++14
